diff --git a/bazel/cython_library.bzl b/bazel/cython_library.bzl
index dc2ef7a890..bd2617a443 100644
--- a/bazel/cython_library.bzl
+++ b/bazel/cython_library.bzl
@@ -60,7 +60,7 @@ def pyx_library(name, deps = [], py_deps = [], srcs = [], **kwargs):
             # Optionally use PYTHON_BIN_PATH on Linux platforms so that python 3
             # works. Windows has issues with cython_binary so skip PYTHON_BIN_PATH.
             cmd =
-                "PYTHONHASHSEED=0 $(location @cython//:cython_binary) --cplus $(SRCS) --output-file $(OUTS)",
+                "PYTHONHASHSEED=0 $(location @cython//:cython_binary) --cplus $(SRCS) --output-file $(OUTS) 2>/dev/null",
             tools = ["@cython//:cython_binary"] + pxd_srcs,
         )
 
