diff --git a/pycross/private/tools/wheel_builder.py b/pycross/private/tools/wheel_builder.py
index c798ce3..1b5437f 100644
--- a/pycross/private/tools/wheel_builder.py
+++ b/pycross/private/tools/wheel_builder.py
@@ -10,6 +10,7 @@ import tarfile
 import tempfile
 import textwrap
 import traceback
+import warnings
 import zipfile
 from pathlib import Path
 from typing import Any
@@ -389,14 +390,16 @@ def link_native_libraries(lib_dir: Path, libraries: List[Path]) -> None:


 def extract_sdist(sdist_path: Path, sdist_dir: Path) -> Path:
-    if sdist_path.name.endswith(".tar.gz"):
-        with tarfile.open(sdist_path, "r") as f:
-            f.extractall(sdist_dir)
-    elif sdist_path.name.endswith(".zip"):
-        with zipfile.ZipFile(sdist_path, "r") as f:
-            f.extractall(sdist_dir)
-    else:
-        _error(f"Unsupported sdist format: {sdist_path}")
+    with warnings.catch_warnings():
+        warnings.simplefilter("ignore")
+        if sdist_path.name.endswith(".tar.gz"):
+            with tarfile.open(sdist_path, "r") as f:
+                f.extractall(sdist_dir)
+        elif sdist_path.name.endswith(".zip"):
+            with zipfile.ZipFile(sdist_path, "r") as f:
+                f.extractall(sdist_dir)
+        else:
+            _error(f"Unsupported sdist format: {sdist_path}")

     # After extraction, there should be a `packageName-version` directory
     (extracted_dir,) = sdist_dir.glob("*")
