diff --git a/js/private/js_binary.bzl b/js/private/js_binary.bzl
index ec334193..d40a5865 100644
--- a/js/private/js_binary.bzl
+++ b/js/private/js_binary.bzl
@@ -129,6 +129,10 @@ _ATTRS = {
         and ["Make variable"](https://bazel.build/reference/be/make-variables) substitution if `expand_env` is set to True.
         """,
     ),
+    "env_inherit": attr.string_list(
+        default = [],
+        doc = "Specifies additional environment variables to inherit from the external environment when the test is executed by bazel test.",
+    ),
     "expand_args": attr.bool(
         default = True,
         doc = """Enables [$(location)](https://bazel.build/reference/be/make-variables#predefined_label_variables)
@@ -572,7 +576,11 @@ def _js_binary_impl(ctx):
     )
     runfiles = launcher.runfiles
 
-    providers = []
+    providers = [
+        RunEnvironmentInfo(
+            inherited_environment = getattr(ctx.attr, "env_inherit", []),
+        ),
+    ]
 
     if ctx.attr.testonly and ctx.configuration.coverage_enabled:
         # We have to instruct rule implementers to have this attribute present.
